src/include/miopen/rnn/algorithms/default_algo_utils.hpp

```diff
@@ -0,0 +1,724 @@
+    }
+
+public:
+    static RNNModuleAlgoBase create(const RNNDescriptor& rnnDesc,
+                                    const SeqTensorDescriptor& xDesc,
+                                    const SeqTensorDescriptor& yDesc,
+                                    const TensorDescriptor& hDesc,
+                                    miopenRNNFWDMode_t mode)
+    {
+        auto [max_layers_hid, max_batch_hid, hidden_vec_sz] = miopen::tien<3>(hDesc.GetLengths());
+        auto [max_batch_in, max_seq, input_vec_sz]          = miopen::tien<3>(xDesc.GetLengths());
+
+        assert(max_batch_in <= max_batch_hid);
+
+        auto layers_cnt         = static_cast<int>(rnnDesc.nLayers);
+        const bool is_seq_bidir = rnnDesc.dirMode == miopenRNNbidirection;
+
+        assert(static_cast<size_t>(layers_cnt) * (is_seq_bidir ? 2 : 1) <= max_layers_hid);
+
+        auto gates_cnt = static_cast<int>(rnnDesc.nHiddenTensorsPerLayer);
+
+        // class update req
+        assert(!is_seq_bidir);
+
+        // TODO all size_t
+        GeneralLstmRedBuffer rb_layout = forwardInterimInfoBuilder(rnnDesc, xDesc);
+
+        GeneralLstmTempBuffer workspace_info = backwardInterimInfoBuilder(rnnDesc, xDesc);
+
+        WeightsBufferDescriptor weights_layout =
+            WeightsBufferDescriptor::create(static_cast<int>(input_vec_sz),
+                                            static_cast<int>(hidden_vec_sz),
+                                            layers_cnt,
+                                            rnnDesc.biasMode,
+                                            rnnDesc.inputMode,
+                                            gates_cnt,
+                                            is_seq_bidir);
+
+        BatchController batch_controller = BatchController::Create(xDesc);
+
+        HiddenBuffersDescriptor hidden_hxcx_info{hDesc};
+
+        IOBufferDescriptor x_info{IOBufferDescriptor::build(xDesc)};
+        IOBufferDescriptor y_info{IOBufferDescriptor::build(yDesc)};
+
+        return {rb_layout,
+                workspace_info,
+                weights_layout,
+                hidden_hxcx_info,
+                x_info,
+                y_info,
+                rnnDesc,
+                batch_controller,
+                mode};
+    }
+
+    RNNModuleAlgoBase(RNNModuleAlgoBase&&)      = default;
+    RNNModuleAlgoBase(const RNNModuleAlgoBase&) = default;
+    // RNNModuleAlgoBase(RNNModuleAlgoBase const&) = default;
+
+    RNNModuleAlgoBase(const GeneralLstmRedBuffer& rb_layout,
+                      const GeneralLstmTempBuffer& workspace_info,
+                      const WeightsBufferDescriptor& weights_layout,
+                      const HiddenBuffersDescriptor& hidden_hxcx_info,
+                      const IOBufferDescriptor& x_info,
+                      const IOBufferDescriptor& y_info,
+                      const RNNDescriptor& rnn_desc,
+                      const BatchController& batch_controller,
```

Comment:
```suggestion
    RNNModuleAlgoBase(GeneralLstmRedBuffer&& rb_layout,
                      GeneralLstmTempBuffer&& workspace_info,
                      WeightsBufferDescriptor&& weights_layout,
                      HiddenBuffersDescriptor&& hidden_hxcx_info,
                      IOBufferDescriptor&& x_info,
                      IOBufferDescriptor&& y_info,
                      RNNDescriptor&& rnn_desc,
                      BatchController&& batch_controller,
```
And move everything. There is a bunch of vectors under the hood of descripotrs.
But probably it's not critical, if all the vectors will be replaced with inline vector (which is coming soon).

Reply:
well, yes and no)  all objects are not standard data descriptors. Part of them is array, part need to convert to array, part should be vector just because stored dynamic data size.
RNNDescriptor only ref, because the lifetime of a class cannot be longer than the lifetime of the object it lives in.